File: algorithms/c/222.c
#include <stdio.h>

struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int countNodes(struct TreeNode* root)
{
    if( root == NULL )
        return 0;

    int heightL = 0, heightR = 0;
    struct TreeNode *p;
    p = root;
    while( p != NULL )
    {
        heightL++;
        p = p->left;
    } 

    p = root;
    while( p != NULL )
    {
        heightR++;
        p = p->right;
    }

    if( heightL == heightR )
        return (1<<heightL)-1;
    return 1 + countNodes(root->left)+countNodes(root->right);
}

int main()
{
    return 0;
}

File: algorithms/python/222.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        depth = self.countDepth(root.left)
        if depth == self.countDepthRight(root.left):
            left = 2 ** depth - 1
            return 1 + left + self.countNodes(root.right)
        else:
            right = 2 ** (depth-1) - 1
            return 1 + self.countNodes(root.left) + right

    def countDepth(self, root):
        depth = 0
        while root:
            depth += 1
            root = root.left
        return depth

    def countDepthRight(self, root):
        depth = 0
        while root:
            depth += 1
            root = root.right
        return depth
